Add initial specs v0.1.0 (#1)
sarisboo/shopping_list: specs/v0.1.0.md

# Shopping list 

> We spend a lot of time picking a recipe, testing it  or trying to remember on which website/cookbook it was  and when we are finally done collecting a list of recipes. We still have to list all the ingredients necessary to go grocery shopping in the right amounts. 

Reviewer: ```suggestion
We spend a lot of time picking a recipe, testing it or trying to remember on which website/cookbook it was and when we are finally done collecting a list of recipes, we still have to list all the ingredients necessary to go grocery shopping in the right amounts. 
```

This project will provide the following features:

> - Generate a ‘recipes’ database and feed initial recipes to it.

Reviewer: ```suggestion
- Generate a `recipes` database and feed initial recipes to it.
```

- Create/Update/Delete/Read recipes and their ingredients.
- Rate recipes.
- Randomly suggest a list of recipes for the week.

What this project will provide in a future version:

> - Adding new recipes to the database  based on preferences.

Reviewer: ```suggestion
- Adding new recipes to the database based on preferences.
```

- Once the recipes are approved by the user, generate a shopping list of all the ingredients needed to make them.

We will to create a SQLite database storing our favourite recipes and issue a selection of 10 recipes for the week via one REST API endpoint. The server will be implemented in Python 3.6.5.

## Project organization

### Poetry

We will use [Poetry](https://poetry.eustace.io/) to structure our source code, manage dependencies, control the Python environment. `poetry` also offers a centralized way of executing several scripts like so: `poetry run <script>`.

> We will provide the following script:

Reviewer: ```suggestion
We will provide the following scripts:
```

>  - `api`: defines the functions that will be queries to the database using the sqlite3 module.

Reviewer: I'm not sure I understand what `poetry run api` will do? Will it start the API server?

>  - `run`: runs the previously defined functions to obtain the recipe list and the final ingredient lists.

Reviewer: I'm not sure what it does either. Does it provide a list of recipes for the week? If so you might have misunderstood what an API does. We are not meant to run `poetry` to have this list of recipes, but instead call an API endpoint via HTTP, something like `curl https://localhost:1337/recipes/week_suggestions`.

Also, you mentioned that the ingredient lists would be made for a future version, so you might want to remove that here.

One last thing, the final Poetry command would be something like `poetry run run` which is probably not something that we want as it's a little confusing.

 - `test`: will be testing our script

### Packages

The project will be composed of the following packages:
 - `shopping_list`: Contains the code for the API that will generate the suggested recipes.
 - `tests`: Contain all the unit tests.

## Database

### Schema

#### Recipes
| Column        | Type   | Attributes    | Description |
| ------------- | ------ | ------------- | ----------- |
| id            | `INT`  | `PRIMARY KEY` | recipe id |
| name          | `TEXT` | `NOT NULL`    | name of the recipe|
| rating        | `INT`  | `NOT NULL`    | grade of the recipe (max value is 5)|
| total_time    | `INT`  | `NOT NULL`    | preparation time + cooking time|

> | ingredient_id | `INT`  | `FOREIGN KEY` | foreign key to the ingredients table|

Reviewer: You will have a problem here. If you simply specify one foreign key, you'll then only be able to link 1 recipe with 1 ingredient. You need to find a way to link multiple recipes with multiple ingredients (a "many-to-many" relationship).

Reviewer: You may want to delete that field.

#### Ingredients

| Column        | Type   | Attributes    | Description |
| ------------- | ------ | ------------- | ----------- |
| id            | `INT`  | `PRIMARY KEY` | ingredient id |
| name          | `TEXT` | `NOT NULL`    | name of the ingredient|
| amount_id     | `INT`  | `FOREIGN KEY` | foreign key to the amounts table|

#### Amounts

| Column        | Type   | Attributes    | Description |
| ------------- | ------ | ------------- | ----------- |
| id            | `INT`  | `PRIMARY KEY` | amount id |

> | name          | `INT`  | `NOT NULL`    | quantity  |

Reviewer: Just a tip related to my previous comment, you can use this table to implement the "many-to-many" relationship.

Reviewer: Delete this.

### Seeding

This will be done using SQLite DB Browser because we can perform tasks such as creating databases, searching them and editing the data in a more visual way.

In a future version we will provide a script to do that automatically.

> For a start, we will add about 13 recipes to the database, the number of recipes will grow over time.

Reviewer: 13 is a pretty exact number 😆 Minor, but if you intend to add "exactly" 13 recipes you might want to remove "about" If you are not sure it will be 13, then something like "about a dozen recipes" will make it!.

## API

> The API will only have have one endpoint for now, which will be selecting the recipes randomly and providing the corresponding list of ingredients.

Reviewer: Can you give the URL scheme of the API and that endpoint?

> for this, we will use SQLite3 a built-in python module which allows us to connect to the database.

Reviewer: Seems redundant with the following sentence.

> We will be using the sqlite3 built in library in python to communicate with the database. This library can be used to insert, select, update, and delete data from the database, for example here is how we would implement a simple GET endpoint:

Reviewer: ```suggestion
We will be using the `sqlite3` built-in library in Python to interact with the database. For example here is how we would implement a simple `GET` endpoint:
```

```
import sqlite3

conn = sqlite3.connect('shopping_list.db')

c = conn.cursor()

> def get_emps_by_name(name):

Reviewer: Kesako `get_emps_by_name`?

    c.execute("SELECT * FROM recipes WHERE name=:name", {'name': recipename})
    return c.fetchall()

conn.close()
```
### Recipes suggestion

> For now we will randomly select and generate a list of 10 recipes each the endpoint is solicited.

Reviewer: ```suggestion
For now, we will randomly generate a list of 10 different recipes each time the endpoint is hit.
```

This project will provide the following features:

- Generate a `shopping_list` database and feed initial recipes to it.
- Create/Update/Delete/Read recipes and their ingredients.
- Rate recipes.
- Randomly suggest a list of recipes for the week.

What this project will provide in a future version:

- Adding new recipes to the database based on preferences.
- Once the recipes are approved by the user, generate a shopping list of all the ingredients needed to make them.

We will to create a SQLite database storing our favourite recipes and issue a selection of 10 recipes for the week via one REST API endpoint. The server will be implemented in Python 3.6.5.

## Project organization

### Poetry

We will use [Poetry](https://poetry.eustace.io/) to structure our source code, manage dependencies, control the Python environment. `poetry` also offers a centralized way of executing several scripts like so: `poetry run <script>`.

We will provide the following scripts:

 - `api`: this module will hold the basic api code built with flask. 

>  - `models`: this package will hold our database models and marshmallow schemas.

Reviewer: We are talking about Poetry scripts like `poetry run server`, not packages or modules here.

 - `test`: the script that will be testing our code.

### Packages

The project will be composed of the following packages:
 - `shopping_list`: Contains the code for the API that will generate the suggested recipes.
 - `tests`: Contain all the unit tests.
 - `models`: this package will hold our database models and marshmallow schemas.

> ## Database

Reviewer: ```suggestion

## Database
```

### Schema

#### Recipes
| Column        | Type   | Attributes    | Description |
| ------------- | ------ | ------------- | ----------- |
| id            | `INT`  | `PRIMARY KEY` | recipe id |
| name          | `TEXT` | `NOT NULL`    | name of the recipe|
| rating        | `INT`  | `NOT NULL`    | grade of the recipe (max value is 5)|
| total_time    | `INT`  | `NOT NULL`    | preparation time + cooking time|
| ingredient_id | `INT`  | `FOREIGN KEY` | foreign key to the ingredients table|

#### recipe_ingredient (joining table)

| Column        | Type   | Attributes    | Description |
| ------------- | ------ | ------------- | ----------- |
| recipe_id     |        | `INT`         | `FOREIGN KEY` | to the recipes table |
| ingredient_id |        | `INT`         | `FOREIGN KEY` | to the ingredients table |

> #### Ingredients

Reviewer: Rename to `ingredients` to be consistent.
Remove the additional line breaks.

| Column        | Type   | Attributes    | Description |
| ------------- | ------ | ------------- | ----------- |
| id            | `INT`  | `PRIMARY KEY` | ingredient id |
| name          | `TEXT` | `NOT NULL`    | name of the ingredient|

> | amount_id     | `INT`  | `FOREIGN KEY` | foreign key to the amounts table|

Reviewer: Remove this as you want to store an amount per recipe/ingredient, not per ingredient.

#### ingredient_amount (joining table)

| Column        | Type   | Attributes    | Description |
| ------------- | ------ | ------------- | ----------- |
| ingredient_id            | `INT`  | `FOREIGN KEY` | to the ingredients table |

> | amount_id                | `INT`  | `FOREIGN KEY` | to the recipes table     |

Reviewer: Delete this.

# Shopping List Tech Specs v0.1.0

## Overview
This project is aimed at optimising the grocery shopping process by generating a list of favourite recipes and from that, extracting another list of necessary ingredients and their amounts.

> Our approach will be to create a relational database of our favourite recipes and extract all the information we need using a REST API. The overall time estimate for this project is one week.

Reviewer: You could move this paragraph in the **Approach** section as an overview of the technical solution you will use.

I would also detail a little more how you would technically implement the features that you will list in the following section (cf. comment below).

> ## Goals and Product Requirement

Reviewer: I would rename this section in **Features** to be a little less *formal* and list the main features of this `v0.1.0`.

The features list should be exhaustive so it's clear what will be implemented (and what won't be). You could also refer to this list of features as the *scope* of the project.

Here is an example of the features you could list and describe:
- Seed the DB with initial recipes.
- Create / Update / Delete recipes with their ingredients.
- Rate recipes.
- Get a list of 10 proposed recipes.

We spend a lot of time picking a recipe, testing it  or trying to remember on which website/cookbook it was  and when we are finally done collecting a list of recipes. We still have to list all the ingredients necessary to go grocery shopping in the night amounts. 

> In order to stop wasting time on these monotonous tasks, we propose a way to automate the process by selecting a number of recipes for the week (taking into account the rating criteria, ingredient similarity etc…) to generate an optimised grocery shopping list. 

Reviewer: What about moving the 2 paragraphs above right after the first one as they detail the features rather than the technical solution.

## Approach 

### SQLite DB Browser

> We will use Sqlite DB Browser to store the database with a table listing all our favourite recipes. Each recipe will have a characteristic ID. Another table will be containing all the ingredients necessary for the listed recipes . 

Reviewer: SQLite DB Browser is a software to interact with SQLite. It doesn't *store* the DB, SQLite does. Referring to the list of features above, how will you use SQLite DB Browser?

More generally, I think you could move this to a **Technology** section in order to describe your basic building blocks there.

Ingredients can be linked to one or more recipe so  the final grocery list will have the total number of each ingredient necessary to the realisation of all selected recipes.

> #### Creating Relational Database

Reviewer: In this section, you seem to describe the structure of the DB. It's known as a *schema*. So I would rename this section to **Database Schema**.

To make it a little more readable, you could use [Markdown tables](https://help.github.com/articles/organizing-information-with-tables/) to format this schema.

1. First table: Recipes

	Each column in the table will have a field name as follows:

> 		- recipe_id: primary key field for each recipe

Reviewer: Is there a reason to prepend `id` with the name of the table in SQLite? I think you can simply use `id` to KISS.

		- recipe_name: name of the recipe
		- rating: grade of the recipe (max value is 5)
		- total_time: prep time + cooking time
		
2. Joining table:  
		- recipe_id
		- ingredient_id
		
3. Second Table : Ingredients

> 		- Ingredient_id:  unique id for each ingredient

Reviewer: Same as above.

		- ingredients_name:  name of the ingredient

> ### Python 3.6.5

Reviewer: You could move this too to a **Technology** section.

This project will be written in Python 3.6.5 and we will use SQLite to create a database. We chose SQLite because it comes with python and doesn’t have to be installed as a separate module.

We are going to create a new utility program called `build_database.py` to create a initialise the database. 
We will also create two python modules `config.py` and `models.py` which will be used by   `build_database.py` 

#### The Config Module

We’re going to configure Flask, Connexion, SQLAlchemy, and Marshmallow here.
	- `config.py`:gets the necessary modules imported into the program and configured. This includes Flask, Connexion, SQLAlchemy, and Marshmallow. Because it will be used by both `build_database.py`
	Here is an example of a  `config.py` file:

#### Models Module

The `models.py` module is created to provide the Person and PersonSchema classes .

> 	- `models.py`is the module where you’ll create the Recipe SQLAlchemy and RecipeSchema Marshmallow class definitions described above. This module is dependent on `config.py` for some of the objects created and configured there.

Reviewer: Those 2 sub-sections are probably part of a **Architecture** section, rather than **Python 3.6.5**.

def update(lname, person):
    """
    This function updates an existing person in the people structure
    :param lname:   last name of person to update in the people structure
    :param person:  person to update
    :return:        updated person structure
    """
    # Does the person exist in people?
    if lname in PEOPLE:
        PEOPLE[lname]["fname"] = person.get("fname")
        PEOPLE[lname]["timestamp"] = get_timestamp()

        return PEOPLE[lname]

    # otherwise, nope, that's an error
    else:
        abort(
            404, "Person with last name {lname} not found".format(lname=lname)
        )

def delete(lname):
    """
    This function deletes a person from the people structure
    :param lname:   last name of person to delete
    :return:        200 on successful delete, 404 if not found
    """
    # Does the person to delete exist?
    if lname in PEOPLE:
        del PEOPLE[lname]
        return make_response(
            "{lname} successfully deleted".format(lname=lname), 200
        )

    # Otherwise, nope, person to delete not found
    else:
        abort(
            404, "Person with last name {lname} not found".format(lname=lname)
        )
```

> ### Poetry

Reviewer: You could also move this to a **Technology** section.

Finally, in order to structure the files needed for the shopping_list repository specify the dependencies, control the python environment and tests  as well as run the code we will be using poetry 0.12.10.

## Test Plan 

## Open Questions
- Do I test using a Test Database ?

> - Do I test using mock ?

Reviewer: You can move those question in the Pull Request (PR) itself by editing the initial comment.
